src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py

```diff
@@ -0,0 +1,722 @@
+"""
+
+
+# positional embedding
+class RotaryEmbeddingNP(linen.Module):
+    dim: int
+    max_seq_len_cached: int
+    base: int = 10000
+
+    def setup(self):
+        self.inv_freq = 1.0 / (self.base ** (jnp.arange(0, self.dim, 2).astype(jnp.float32) / self.dim))  # dim
+
+    def __call__(self, x=None, seq_len=None):
+        t = jnp.arange(self.max_seq_len_cached, dtype=self.inv_freq.dtype)
+        freqs = jnp.outer(t, self.inv_freq)
+        emb = jnp.concatenate((freqs, freqs), axis=-1)
+        cos_cached = jnp.expand_dims(jnp.expand_dims(jnp.cos(emb), 0), 0)
+        sin_cached = jnp.expand_dims(jnp.expand_dims(jnp.sin(emb), 0), 0)
+        return cos_cached[:seq_len, ...], sin_cached[:seq_len, ...]
+
+
+def rotate_half(x):
+    x1 = x[..., : x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2 :]
+    return jnp.concatenate((-x2, x1), axis=-1)
+
+
+# q,k are of shape [bs, n_heads, seq_len, head_dim]
+# cos, sin are of shape #[1, 1, seq_len, dim]
+# position_ids is of shape [bs, seq_len]
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids):
+    gather_indices = position_ids[:, :, None, None]  # [bs, seq_len, 1, 1]
+    gather_indices = jnp.repeat(gather_indices, cos.shape[1], axis=1)
+    gather_indices = jnp.repeat(gather_indices, cos.shape[3], axis=3)
+    cos = jnp.take_along_axis(cos.repeat(gather_indices.shape[0], axis=0), gather_indices, axis=2)
+    sin = jnp.take_along_axis(sin.repeat(gather_indices.shape[0], axis=0), gather_indices, axis=2)
+    q_embed = (q * cos) + (rotate_half(q) * sin)
+    k_embed = (k * cos) + (rotate_half(k) * sin)
+    return q_embed, k_embed
+
+
+class FlaxGPTNeoXAttention(linen.Module):
+    config: GPTNeoXConfig
+    dtype: jnp.dtype = jnp.float32
+    causal: bool = True
+    is_cross_attention: bool = False
+
+    def setup(self):
+        config = self.config
+        self.num_attention_heads = config.num_attention_heads
+        self.hidden_size = config.hidden_size
+        self.head_size = self.hidden_size // self.num_attention_heads
+        self.rotary_ndims = int(self.head_size * config.rotary_pct)
+        self.rotary_emb = RotaryEmbeddingNP(
+            dim=self.rotary_ndims, max_seq_len_cached=config.max_position_embeddings, base=config.rotary_emb_base
+        )
+        self.norm_factor = jnp.sqrt(self.head_size)
+        self.query_key_value = linen.Dense(
+            3 * config.hidden_size,
+            dtype=self.dtype,
+            kernel_init=jax.nn.initializers.normal(self.config.initializer_range),
+        )
+        self.dense = linen.Dense(
+            config.hidden_size,
+            dtype=self.dtype,
+            kernel_init=jax.nn.initializers.normal(self.config.initializer_range),
+        )
+
+        self.causal_mask = make_causal_mask(jnp.ones((1, config.max_position_embeddings), dtype="bool"), dtype="bool")
+
+    @linen.compact
+    def _concatenate_to_cache(self, key, value, query, attention_mask):
```

Comment:
Let's try and use `# Copied from...` everywhere we re-use existing modelling code

```suggestion
# Copied from transformers.models.gpt_neo.modeling_flax_gpt_neo.FlaxGPTNeoSelfAttention._concatenate_to_cache
def _concatenate_to_cache(self, key, value, query, attention_mask):
```